File: Data Structures & Algorithms/add-two-numbers-ii/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    stack<int> parse(ListNode* node) {
        stack<int> s;
        while (node) {
            s.push(node->val);
            node = node->next;
        }
        return s;
    }

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        stack<int> s1 = parse(l1);
        stack<int> s2 = parse(l2);
        int carry = 0;
        ListNode* next = nullptr;

        while (!s1.empty() || !s2.empty()) {
            int num = carry;
            if (!s1.empty()) {
                num += s1.top();
                s1.pop();
            }
            if (!s2.empty()) {
                num += s2.top();
                s2.pop();
            }

            ListNode* node = new ListNode(num % 10, next);
            next = node;
            carry = num / 10;
        }

        if (carry) {
            ListNode* node = new ListNode(1, next);
            return node;
        } else {
            return next;
        }
    }
};

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        h = []

        for i in range(len(lists)):
            node = lists[i]
            heapq.heappush(h, (node.val, i))
        
        head = None
        prev = None

        while h:
            _, i = heapq.heappop(h)

            node = lists[i]

            if head is None:
                head = node

            if prev:
                prev.next = node

            prev = node

            if node.next:
                heapq.heappush(h, (node.next.val, i))
                lists[i] = lists[i].next

        if prev:
            prev.next = None

        return head